### ISSUE TITLE
Study performance impact of sorting tracks

### ISSUE THREAD

**Issue by amandalund** [celeritas-project/celeritas#675]
Some optimizations to try:
  - Sort tracks by `TrackStatus` once per stepping loop iteration (and still use masking)
  - Sort tracks by `TrackStatus` and action ID before launching each action -- this way we can reduce the grid size by knowing which threads to apply a given action kernel to
  - Try sorting by particle type, energy, etc. in addition to action ID

**Comment by sethrj** [comment]
[merged this into the checklist above]

For the first step (at least) the second case should result in identical answers to the current implementation, so it's also a good check of the index array working. (I *think* it might give the correct answers for subsequent iterations too, but it depends on how we initialize primaries in the empty track slots.)

**Comment by esseivaju** [comment]
Shuffling of track slots indirection is done in 040f9a6. I'm using a RngEngine from the stdlib as the ones in Celeritas are designed to be specific to a track, expecting a `TrackSlotId` as constructor argument. I suppose I could create a view for an arbitrary track in resize, call `make_rng_engine` and pass it as an argument to `shuffle_track_slots` but it looks like a hack.

**Comment by esseivaju** [comment]
If I'm correct the `resize` function for state is only called once when we instantiate the stepper - at initialization. Since we want to sort the tracks every stepping iteration, should we do that in `ActionSequence::execute` or do you think that adding an action would be better?

**Comment by sethrj** [comment]
Resize should just be for assigning then; I think the quickest thing to do would be to add the sort inside ActionSequence. In fact it may even be reasonable to add different strategies as a runtime option in there...

**Comment by sethrj** [comment]
@esseivaju Given the result from https://github.com/celeritas-project/celeritas/commit/040f9a66522af657111976745d90c6ff91d3b513 that `thrust::shuffle` isn't doing the shuffle on-device, and since the shuffling is just to study whether there's an effect for out-of-order memory access, maybe we should just shuffle on the host only at the beginning? I think also it might be good for testing purposes to put these different attempts on the develop branch with a front end option to switch, rather than having to use different builds to differentiate "shuffled" from "unsorted" from "sorted".

**Comment by esseivaju** [comment]
#701 implements config options for sorting tracks

**Comment by esseivaju** [comment]
@sethrj As discussed in #701 I did a quick profiling comparison of the shuffled / unsorted tracks.

## Setup

**System**: Perlmutter GPU (A100)
**Problem**: cms2018+MSC+Field
**Celeritas version**: 2b8916b

I profiled 10 executions of `pre_step_kernel`, starting at step 346 (peak in # active tracks). 

## Overall statistics

There is a big difference in the total execution time of the kernel. When shuffled, it is about 2x slower (1.7ms vs 800us). The compute throughput is also half of the unsorted version:

![image](https://user-images.githubusercontent.com/7202402/228377836-eae2768a-3acd-40af-95f7-b5bbcbfb200f.png)

![image](https://user-images.githubusercontent.com/7202402/228378022-3cd024d8-9963-41ff-a4e6-0a5db3cba664.png)


## Memory workload

Memory access pattern is significantly worse, as expected. Cache hit rate is significantly lower, and we're transferring twice as much data between Global -> L2 and L2 -> L1

![image](https://user-images.githubusercontent.com/7202402/228378609-eec752fe-62c3-49dc-b80f-f641a5cfff0a.png)

## Warps state

Given that the compute throughput is much lower and that we're doing much more memory transfer, we can expect threads to be waiting on memory. This is definitely the case, the number of cycles per instruction is doubled when we shuffle tracks (42 cycles/instructions vs 21 on average). 30 cycles are spent waiting on memory and 4 waiting on L1 instruction queue when we shuffle tracks vs 10 and 1, respectively:

  
![image](https://user-images.githubusercontent.com/7202402/228380384-e6e0e9e0-bd78-4cd3-bf83-c53d32b00dde.png)

**Comment by esseivaju** [comment]
I also took a quick look at the `along_step_uniform_msc_kernel`. Overall it's similar to the `pre_step_kernel`, performances are worse when shuffled (total execution time is 70ms vs 85ms, ~20% slower when shuffled). Of particular interest, it seems that the L2 cache utilization is quite a bit worse: 

![image](https://user-images.githubusercontent.com/7202402/228388158-b4576a40-158f-4e82-aca6-db64ab63fa18.png)

Looking at L1 load to L2, we have a much lower hit rate, we need to issue way more requests and transfer 3x more data from global memory to L2.

**Comment by sethrj** [comment]
That's really interesting. Is this then an effect of coalescing? From our earlier studies I thought that ended up being almost irrelevant. We could *at least* try converting the 'status' member of `SimTrackState` from AOS to SOA; that guy is used everywhere and (because it's a single byte that ends up padded in a larger struct) it would certainly reduce memory bandwidth.

**Comment by esseivaju** [comment]
Yes, more requests would indicate that memory access are not coalesced as well but I'm not sure it's the only issue. Even in the default - unsorted - case we're not really doing coalesced memory access. In fact a L2 load access on average 1.3 sectors out of the 4 of a cache line. When shuffled, we access of average 1.1 sector so the increase in number of requests seems disproportionate.

I can have a look at moving SimTrackState to a SOA but I'll be quite busy this week and next week (preparing CHEP slides and another deadline coming soon...) so not sure how quick I'll progress on that.

**Comment by sethrj** [comment]
Yeah, I'm having trouble understanding why it's so much worse... did you also happen to check that the number of steps are the same?

**Comment by esseivaju** [comment]
yes, same number of steps, same number of alive and active tracks at each step

**Comment by esseivaju** [comment]
I have ported `SimTrackView` state data to a SoA in #707. 

For the along_step kernel, this doesn't really make a difference. Execution time for the kernel is the same. This slightly reduce the data transferred between cache levels and the # of requests:

![image](https://user-images.githubusercontent.com/7202402/228695922-f6245190-194f-44f2-af64-5db40810a43e.png)

For the pre_step_kernel, this help a lot more, total time is 20% faster. Unfortunately that's not the kernel we want to optimize ☹️ . Data transferred  between L1-L2 is much lower, particularly store requests are halved: 

![image](https://user-images.githubusercontent.com/7202402/228696871-66fc2f8d-3155-4b46-b6cd-0582bc9f2100.png)

**Comment by sethrj** [comment]
Wow, 20% faster is extreme even though that's not as important. Are these before-and-after results both with unsorted tracks?

**Comment by esseivaju** [comment]
Yes, I only compared unsorted. The total time for one execution of pre_step goes from ~810us to ~640us

**Comment by esseivaju** [comment]
@sethrj do we currently have a correspondence between how particles / tracks are ordered in device memory and any of their properties or is it totally arbitrary?

**Comment by sethrj** [comment]
@esseivaju The only correspondence should be on the `TrackSlotId`. There's currently no cooperation between GPU threads and track slots.

**Comment by esseivaju** [comment]
So if we're sorting tracks by e.g. ParticleId or energy, so that contiguous threads access similar tracks, I suppose this will help thread divergence as similar tracks are likely to follow a similar code path? Memory access would still be scattered though, we'd need to actually reorder the data in `PhysicsStateData` and `ParticleStateData`

**Comment by sethrj** [comment]
Exactly. The reordering would help divergence but likely hurt memory access patterns.

**Comment by sethrj** [comment]
Although I think the justification for doing so is "very little we do is coalesced anyway [random access patterns] so it probably doesn't matter"?

**Comment by esseivaju** [comment]
Do you think it doesn't matter at all? I'd expect that if tracks that are likely to follow the same code path have their data in contiguous memory and are processed by neighboring threads I'd assume coalesced access would occur more frequently

**Comment by sethrj** [comment]
I think that was the conclusion that @tmdelellis et al had from Shift. It might make sense for us to sort only on a couple of kernels, and we might see more of an effect during the tail end where fewer and fewer tracks are active. I think we'll need to experiment to see... it might make sense to do it on only a subset of kernels too. The along-step kernel might see more of a difference whenever we implement different kernels for photons and charged particles.

**Comment by esseivaju** [comment]
A few comments after running cms2018+msc+field-vecgeom on Perlmutter w/wo sorting by action id:

- walltime is ~8% shorter when sorting by action id so overall it seems worth it
- the along_step kernel - tested when track slots are full - is faster (from 10% to 20%)
  - The main reason is that warp divergence is lower so we have fewer cycles per call
- we're transferring more bytes between each cache levels and device memory, with fewer cycles this means that the arithmetic intensity of the kernel is significantly lower. We're memory bound, maybe transforming more state struct to SoA could help? `ParticleStateData` / `PhysicsStateData` ?
- Sorting takes quite a bit of time; this is the 3rd slowest action after along_step and geo_boundaries. It is still significantly shorter than along_step, only ~5% of along_step total time.

**Comment by sethrj** [comment]
Note on profiling (https://github.com/celeritas-project/regression/blob/perlmutter-profiling/run-problems.py):
```python
[
            "ncu",
            "-f",
            f"--export={filename}",
            f"--set={ncu_set}",
            "--launch-skip=345",
            "--launch-count=10",
            f"--kernel-name={kernel_name}",
            build / "app" / "demo-loop",
            "-"
]
```

**Comment by esseivaju** [comment]
@sethrj Using NVTX to get a more detailed breakdown of where the time is spent during the SortTracks Action. There is some memory allocation but the time spend in cudaMalloc can vary a lot between steps. The image below shows a step with the maximum number of tracks. The sort action is broken down in 3 substeps: sorting, counting and back-filling on CPU. Most of the time is spent on the sorting step. The biggest gap in GPU utilization comes while we're waiting to deallocate some memory (probably internal to thrust) before stating to count tracks. 

![image](https://github.com/celeritas-project/celeritas/assets/7202402/29eadcfc-d62a-49c9-be4a-653cb8dccfa8)

I've seen some cases where the call to cudaMalloc takes an abnormally long time, for example here it takes 13ms

![image](https://github.com/celeritas-project/celeritas/assets/7202402/f2ae81eb-4b69-4418-998b-daf63589ce8b)

Note that I'm running this on a Perlmutter login node as I'm getting an error writing the profiling output on a compute node... so there might be some noise as people use the node.

**Comment by esseivaju** [comment]
@sethrj @amandalund Test run of the cms2018 problem, that's the timing we get for sorting now, the overhead is way smaller than before but unfortunately doesn't help much reduce the along-step time. I also have Nsight systems profiling output if you want to take a look at them. Most of the time in sorting is spent [copying](https://github.com/celeritas-project/celeritas/blob/1596c767cdadc620c258ea3dd85660881ecdc216/src/celeritas/track/detail/TrackSortUtils.cu#L253-L254) the action count back to the host which can probably be improved by using pinned memory there. I'll try to get everything compiled on a system with AMD gpus and see how it does there.
```
esseivaj@login27:/global/cfs/cdirs/atlas/esseivaj/devel/regression> jq '.result["runner"]["time"]["actions"]' results/perlmutter/cms2018+field+msc-vecgeom-gpu/0.json 
{
  "along-step-neutral": 12.297673017999973,
  "along-step-uniform-msc": 47.38649688000021,
  "annihil-2-gamma": 0.24790105500000081,
  "brems-rel": 0.30715188400000004,
  "brems-sb": 0.48315196500000274,
  "conv-bethe-heitler": 0.3389869929999983,
  "extend-from-primaries": 0.020372580000000362,
  "extend-from-secondaries": 1.4460137470000025,
  "geo-boundary": 6.019717414000041,
  "initialize-tracks": 0.23472176200000364,
  "ioni-moller-bhabha": 0.386410014000001,
  "photoel-livermore": 0.8025920850000005,
  "physics-discrete-select": 0.49531618900000013,
  "pre-step": 1.3510001069999988,
  "scat-klein-nishina": 0.31279201600000234,
  "scat-rayleigh": 0.30189068399999996
}
esseivaj@login27:/global/cfs/cdirs/atlas/esseivaj/devel/regression> jq '.result["runner"]["time"]["actions"]' results/perlmutter/cms2018+field+msc-vecgeom-gpu/0.sorted_alongstep.json 
{
  "along-step-neutral": 10.781980414999992,
  "along-step-uniform-msc": 51.55514914900011,
  "annihil-2-gamma": 0.2789525929999985,
  "brems-rel": 0.3157409370000015,
  "brems-sb": 0.44291699400000073,
  "conv-bethe-heitler": 0.3319529289999995,
  "extend-from-primaries": 0.020277015999999815,
  "extend-from-secondaries": 1.442664754999989,
  "geo-boundary": 5.0736732759999805,
  "initialize-tracks": 0.23476415100000597,
  "ioni-moller-bhabha": 0.37169614200000206,
  "photoel-livermore": 0.8188562790000008,
  "physics-discrete-select": 0.516837046999994,
  "pre-step": 2.0472347579999997,
  "scat-klein-nishina": 0.3090604970000005,
  "scat-rayleigh": 0.30845938800000167,
  "sort-tracks-along-step": 1.9713972230000063
}
```

**Comment by sethrj** [comment]
Thanks @esseivaju . Given the issues with VecGeom virtual function performance, can you also run testem3 with orange vs VecGeom?

**Comment by esseivaju** [comment]
yeah I'll try that

**Comment by esseivaju** [comment]
I have timing [here](https://github.com/celeritas-project/regression/tree/8d97b3e9f90c8fbe4384b18e1adffb76bb2b6c98/results/perlmutter) of orange vs. vecgeom w/wo sorting before the alongstep.

**Comment by sethrj** [comment]
Bleh so it slows down ORANGE substantially....

**Comment by esseivaju** [comment]
Running the nsys profiler after #965, I see the same performance with CUDA as shown in the timing for cms2018+msc+field:

```json
{
  "along-step-neutral": 10.755985781999982,
  "along-step-uniform-msc": 51.47460398100008,
  "annihil-2-gamma": 0.2794058539999981,
  "brems-rel": 0.31380840300000085,
  "brems-sb": 0.442074241000001,
  "conv-bethe-heitler": 0.3307707130000006,
  "extend-from-primaries": 0.019941909999999077,
  "extend-from-secondaries": 1.4611483679999975,
  "geo-boundary": 5.071120742999985,
  "initialize-tracks": 0.2337846580000036,
  "ioni-moller-bhabha": 0.3717192530000025,
  "photoel-livermore": 0.8195309779999975,
  "physics-discrete-select": 0.5177204220000012,
  "pre-step": 2.065046826999991,
  "scat-klein-nishina": 0.30964679499999864,
  "scat-rayleigh": 0.307409995000001,
  "sort-tracks-along-step": 1.9366414099999965
}
```

The overhead of sorting is small; about ~1.5% of the total time on sorting kernels. On CPU, we no longer see the slow memcpy to non-pinned memory but since we have to synchronize at the end of the sort action anyway this doesn't matter. The issue is that the uniform along-step kernel is consistently slower when sorting tracks 😓 

The first two screenshots illustrate the effect: while the GPU is still working on the pre-step kernel, the CPU gets ahead and starts scheduling sorting kernels. Once everything is scheduled and we need results back on the CPU, we have to wait for the GPU to catch up, whether it's because the memcpy is not asynchronous or the explicit stream synchronization. The synchronization could be removed if we don't want to adjust the kernel grid size, maybe make that optional? Even then, I don't think it would help since synchronization is necessary at the end of each step, so the CPU can't get far ahead.

<img width="1071" alt="image" src="https://github.com/celeritas-project/celeritas/assets/7202402/47cade80-8cc6-4eff-ad52-92deb4cd5691">

<img width="1554" alt="image" src="https://github.com/celeritas-project/celeritas/assets/7202402/811c7a6f-25a2-4874-ae00-231b99e21962">

The overhead seems very close to minimal now, I don't think we can reduce it much more.

**Comment by sethrj** [comment]
Thanks @esseivaju , you've done a great job reducing the possible sources of overhead for sorting. It's quite possible that the performance drop from sorting is due to vecgeom divergence. Could you try running testem3 with orange to see how that compares?

**Comment by sethrj** [comment]
Another thing that I think could show the benefit of the smaller grid sizes (which we have due to sorting) is multithread mode, where we have even more kernels competing for GPU time, and possibly even more latency hiding. Could you try disabling `merge_events` and running with multithreaded through celer-sim? Additionally, running tilecal with the track sorting could show us the possible "real life" performance difference.

**Comment by esseivaju** [comment]
I've done measurements with testem3/orange, just before fixing the memcpyasync (see previous comments) so I'm not expecting any difference from what we already have but I could give it another go.

Yes, having smaller grids should definitely help in multithreaded mode where we can fit more kernels on the GPU. I'll try running multithreaded celer-sim / tilecal with sorting enabled.

**Comment by sethrj** [comment]
Sorry, forgot about the previous ORANGE result! I agree it shouldn't make much difference.

**Comment by esseivaju** [comment]
I've tested both cms2018+msc+field and Tilecal with multiple threads. 

| | cms2018 16t | Tilecal 32t |
|--------|--------|--------|
| sort | 91s | 15s |
| no sort | 106s | 13.8s | 

Sorting consistently improves the performance for the cms2018 problem but decreases the performance with the tilecal. This is explained by the fact that in the Tilecal, there is no magnetic field and the geometry is simpler, the propagation kernel takes ~50 $\mu s$ on average, and the radix sort kernel is a significant overhead in that case, about 11% of the GPU time and the propagation takes only ~10%. In the cms2018 problem, the propagation takes 65% of GPU time and 4.7ms on average so the overhead isn't as significant, and reducing the grid size of this large kernel is more important. I have the profiling data if you're interested in taking a look at them.
